src/parsing/prscParser.ts

```diff
@@ -0,0 +1,2795 @@
+		or([
+			map(preceded(surrounded(token(':='), whitespace), varValue), (x) => ['varValue', x]),
+			map(
+				precededMultiple(
+					[whitespacePlus, token('external')],
+					optional(preceded(surrounded(token(':='), whitespace), varDefaultValue))
+				),
+				(_) => ['external']
+			),
+		]),
+		(itemTypePart, val) =>
+			[
+				'contextItemDecl',
+				...(itemTypePart ? [['contextItemType', itemTypePart]] : []),
+				val,
+			] as IAST
+	);
+
+	const prolog: Parser<IAST> = then(
+		star(
+			followed(
+				or([defaultNamespaceDecl, setter, namespaceDecl, importExpr]),
+				surrounded(separator, whitespace)
+			)
+		),
+		star(
+			followed(
+				or([contextItemDecl, annotatedDecl, optionDecl]),
+				surrounded(separator, whitespace)
+			)
+		),
+		(moduleSettings, declarations) =>
+			moduleSettings.length === 0 && declarations.length === 0
+				? null
+				: ['prolog', ...moduleSettings, ...declarations]
+	);
+
+	const moduleDecl: Parser<IAST> = precededMultiple(
+		[token('module'), whitespacePlus, token('namespace'), whitespacePlus],
+		then(
+			followed(ncName, surrounded(token('='), whitespace)),
+			followed(uriLiteral, surrounded(separator, whitespace)),
+			(prefixPart, uri) => ['moduleDecl', ['prefix', prefixPart], ['uri', uri]]
+		)
+	);
+
+	const libraryModule: Parser<IAST> = then(
+		moduleDecl,
+		preceded(whitespace, prolog),
+		(moduleDeclPart, prologPart) =>
+			['libraryModule', moduleDeclPart, ...(prologPart ? [prologPart] : [])] as IAST
+	);
+
+	const mainModule: Parser<IAST> = then(
+		prolog,
+		preceded(whitespace, queryBody),
+		(prologPart, body) => ['mainModule', ...(prologPart ? [prologPart] : []), body]
+	);
+
+	const versionDeclaration: Parser<IAST> = map(
+		precededMultiple(
+			[token('xquery'), whitespace],
+			followed(
+				or([
+					then(
+						preceded(token('encoding'), whitespacePlus),
+						stringLiteral,
+						(encoding) => ['encoding', encoding] as IAST | IAST[]
+					),
+					then(
+						precededMultiple([token('version'), whitespacePlus], stringLiteral),
+						optional(
+							precededMultiple(
+								[whitespacePlus, token('encoding'), whitespacePlus],
+								stringLiteral
+							)
+						),
+						(version, encoding) =>
+							[
+								['version', version],
+								...(encoding ? [['encoding', encoding]] : []),
+							] as IAST[]
+					),
+				]),
+				preceded(whitespace, separator)
+			)
+		),
+		(x) => ['versionDecl', ...x]
+	);
+
+	const module: Parser<IAST> = then(
+		optional(surrounded(versionDeclaration, whitespace)),
+		or([libraryModule, mainModule]),
+		(versionDecl, modulePart) =>
+			['module', ...(versionDecl ? [versionDecl] : []), ...[modulePart]] as IAST
+	);
+
+	return surrounded(module, whitespace);
+}
+
+export function parseUsingPrsc(
+	xpath: string,
+	options: { outputDebugInfo: boolean; xquery: boolean },
+	shouldConsumeAll: boolean = true
+): ParseResult<IAST> {
+	return (shouldConsumeAll ? complete(generateParser(options)) : generateParser(options))(
```

Comment:
Besides the TODOs in this file, lets try to not call this `generateParser` runTime. there are only going to be two or three unique versions anyway: the XPath one, the XQuery one and an XQuery Update Facility one.

also, the shouldConsumeAll is I think not really something we'll ever set to `false`, right?

Reply:
The `shoudConsumeAll` was an option used during debugging which can indeed be removed. As for calling `generateParser` at runtime, I have not yet found a way to easily clear all the caches of the specific parsing functions.